subsys/net/ip/tcp2.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2018-2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Transmission Control Protocol (TCP)
+ *
+ * - net_tcp_get() is called by net_context_get(AF_INET, SOCK_STREAM,
+     IPPROTO_TCP, ...) and creates struct tcp for the net_context
+ * - net_tcp_listen()/net_tcp_accept() listen/accept
+ * - At the reception of SYN on the listening net_context, a new pair
+ *   of net_context/struct tcp registers a new net_conn handle
+ *   with the tcp_pkt_received() as a callback
+ * - net_tcp_queue() queues the data for the transmission
+ * - The incoming data is delivered up through the context->recv_cb
+ * - net_tcp_put() closes the connection
+ *
+ * NOTE: The present API is provided in order to make the integration
+ *       into the ip stack and the socket layer less intrusive.
+ *
+ *       Semantically cleaner use is possible (and might be exposed),
+ *       look into the unit test tests/net/tcp2 for insights.
+ */
+
+#ifndef TCP2_H
+#define TCP2_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+
+struct tcp; /* TODO: drop this forward declaration */
+
+/**
+ * @brief Create a TCP connecton for the net_context
+ *
+ * @param context Network context
+ *
+ * @return 0 on success, < 0 on error
+ */
+int net_tcp_get(struct net_context *context);
+
+/**
+ * @brief Close and delete the TCP connecton for the net_context
+ *
+ * @param context Network context
+ *
+ * @return 0 on success, < 0 on error
+ */
+int net_tcp_put(struct net_context *context);
+
+/* TODO: Clarify what happens if the ref count goes to 0 */
+/**
+ * @brief Unref a TCP connecton
+ *
+ * @param context Network context
+ *
+ * @return 0 if successful, < 0 on error
+ */
+int net_tcp_unref(struct net_context *context);
+/* TODO: Merge net_tcp_unref() and net_tcp_put() */
+
+/**
+ * @brief Register an accept callback
```

Comment:
Docstring appears to be (erroneous) copy-paste from net_tcp_accept().

Reply:
@pfalcon 

Thanks, fixed.